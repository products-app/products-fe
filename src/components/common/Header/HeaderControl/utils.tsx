// TO DO: refactor icon, must be string and svg needs async load
import { SignOut, Package } from 'phosphor-react'

export const userMenuOptions: app.DropdownItem[] = [
  {
    label: 'Meus pedidos',
    link: '/user/orders',
    icon: <Package className="text-gray200 text-lg" />,
  },
  {
    label: 'Fazer logout',
    link: '/account/logout',
    icon: <SignOut className="text-gray200 text-lg" />,
  },
]
