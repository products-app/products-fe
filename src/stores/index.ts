import { create } from 'zustand'
import { persist, createJSONStorage } from 'zustand/middleware'
import { createUserSlice } from './slices/userSlice'
import { createCartSlice } from './slices/cartSlice'
import { createAppSlice } from './slices/appSlice'

export const useStore = create(
  persist<app.ICart & app.IUser & app.IAppControl>(
    (set, get) => ({
      ...createUserSlice(set),
      ...createCartSlice(set, get),
      ...createAppSlice(set),
    }),
    {
      name: 'store',
      storage: createJSONStorage(() => sessionStorage),
    },
  ),
)
