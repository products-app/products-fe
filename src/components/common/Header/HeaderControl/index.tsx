import { Text } from '@lebernardo/react'
import { ShoppingCartSimple, UserCircle } from 'phosphor-react'
import Dropdown from '@/components/common/Dropdown'
import { useHeader } from '../hooks'
import { HeaderControlContainer, BadgeContainer } from './styles'
import { userMenuOptions } from './utils'

const HeaderControl = () => {
  const { toggleSidebar, redirectToAccount, user, totalCartItems } = useHeader()

  return (
    <HeaderControlContainer>
      {user ? (
        <>
          {user.name && (
            <Text>
              Ol√°, <span>{user.name}</span>
            </Text>
          )}
          <Dropdown items={userMenuOptions} />
        </>
      ) : (
        <button aria-label="button-redirect-login" onClick={redirectToAccount}>
          <UserCircle />
        </button>
      )}

      <button
        aria-label="button-toggle-sidebar"
        data-cy="button-cart"
        onClick={toggleSidebar}
      >
        <ShoppingCartSimple />
        <BadgeContainer>{totalCartItems}</BadgeContainer>
      </button>
    </HeaderControlContainer>
  )
}

export default HeaderControl
