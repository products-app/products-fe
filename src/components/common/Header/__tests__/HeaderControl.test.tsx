import { render, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { useStore } from '@/stores'
import HeaderControl from '../HeaderControl'

const userMock = {
  id: 1,
  name: 'Pedro',
  token: 'eyJhbGciOiJSUzI1NiIsIng1dCI6Ij',
}

const cartItemsMock = [
  {
    id: 1,
    name: 'Product 1',
    price: 29.9,
    stock: 10,
    quantity: 3,
  },
]

const initialState = useStore.getState()

describe('Header Control component', () => {
  afterEach(() => {
    jest.restoreAllMocks()
  })

  test('should render correctly', () => {
    const { container } = render(<HeaderControl />)

    expect(screen.getAllByRole('button')[0]).toBeDefined()
    expect(container).toMatchSnapshot()
  })

  test('should render correctly when user is logged', () => {
    useStore.setState(
      { ...initialState, user: userMock, items: cartItemsMock },
      true,
    )
    const { container } = render(<HeaderControl />)

    expect(screen.getAllByRole('button')[0]).toBeDefined()
    expect(container).toMatchSnapshot()
  })

  test('should click to open the cart sidebar', async () => {
    render(<HeaderControl />)

    const handleClick = jest.fn()
    const toggleSidebarButton = screen.getByRole('button', {
      name: 'button-toggle-sidebar',
    })

    toggleSidebarButton.addEventListener('click', handleClick, false)
    await userEvent.click(toggleSidebarButton)

    expect(handleClick).toHaveBeenCalled()
  })

  test('should redirect to login page', async () => {
    useStore.setState(
      { ...initialState, user: undefined, items: undefined },
      true,
    )
    render(<HeaderControl />)

    const handleClick = jest.fn()
    const toggleSidebarButton = screen.getByRole('button', {
      name: 'button-redirect-login',
    })

    toggleSidebarButton.addEventListener('click', handleClick, false)
    await userEvent.click(toggleSidebarButton)

    expect(handleClick).toHaveBeenCalled()
  })
})
