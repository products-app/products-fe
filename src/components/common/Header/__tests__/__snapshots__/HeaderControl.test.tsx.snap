// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header Control component should render correctly 1`] = `
<div>
  <div
    class="c-cGenGQ"
  >
    <button
      aria-label="button-redirect-login"
    >
      <svg
        fill="currentColor"
        height="1em"
        viewBox="0 0 256 256"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <rect
          fill="none"
          height="256"
          width="256"
        />
        <circle
          cx="128"
          cy="128"
          fill="none"
          r="96"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="16"
        />
        <circle
          cx="128"
          cy="120"
          fill="none"
          r="40"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="16"
        />
        <path
          d="M63.8,199.4a72,72,0,0,1,128.4,0"
          fill="none"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="16"
        />
      </svg>
    </button>
    <button
      aria-label="button-toggle-sidebar"
      data-cy="button-cart"
    >
      <svg
        fill="currentColor"
        height="1em"
        viewBox="0 0 256 256"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <rect
          fill="none"
          height="256"
          width="256"
        />
        <circle
          cx="80"
          cy="216"
          r="16"
        />
        <circle
          cx="184"
          cy="216"
          r="16"
        />
        <path
          d="M42.3,72H221.7l-26.4,92.4A15.9,15.9,0,0,1,179.9,176H84.1a15.9,15.9,0,0,1-15.4-11.6L32.5,37.8A8,8,0,0,0,24.8,32H8"
          fill="none"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="16"
        />
      </svg>
      <div
        class="c-egqsFx"
      >
        0
      </div>
    </button>
  </div>
</div>
`;

exports[`Header Control component should render correctly when user is logged 1`] = `
<div>
  <div
    class="c-cGenGQ"
  >
    <p
      class="c-dgMvMW c-dgMvMW-kWyXas-size-md"
    >
      Ol√°, 
      <span>
        Pedro
      </span>
    </p>
    <div
      class="c-lXviy"
    >
      <button
        class="c-khCkZv"
      >
        <svg
          fill="currentColor"
          height="1em"
          viewBox="0 0 256 256"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <rect
            fill="none"
            height="256"
            width="256"
          />
          <circle
            cx="128"
            cy="128"
            fill="none"
            r="96"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="16"
          />
          <circle
            cx="128"
            cy="120"
            fill="none"
            r="40"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="16"
          />
          <path
            d="M63.8,199.4a72,72,0,0,1,128.4,0"
            fill="none"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="16"
          />
        </svg>
      </button>
      <div
        class="c-fwKBLg"
      >
        <div
          class="c-biAswx"
        >
          <a
            href="/user/orders"
          >
            <svg
              class="text-gray200 text-lg"
              fill="currentColor"
              height="1em"
              viewBox="0 0 256 256"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <rect
                fill="none"
                height="256"
                width="256"
              />
              <path
                d="M224,177.3V78.7a8.1,8.1,0,0,0-4.1-7l-88-49.5a7.8,7.8,0,0,0-7.8,0l-88,49.5a8.1,8.1,0,0,0-4.1,7v98.6a8.1,8.1,0,0,0,4.1,7l88,49.5a7.8,7.8,0,0,0,7.8,0l88-49.5A8.1,8.1,0,0,0,224,177.3Z"
                fill="none"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="16"
              />
              <polyline
                fill="none"
                points="177 152.5 177 100.5 80 47"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="16"
              />
              <polyline
                fill="none"
                points="222.9 74.6 128.9 128 33.1 74.6"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="16"
              />
              <line
                fill="none"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="16"
                x1="128.9"
                x2="128"
                y1="128"
                y2="234.8"
              />
            </svg>
             
            Meus pedidos
          </a>
          <a
            href="/account/logout"
          >
            <svg
              class="text-gray200 text-lg"
              fill="currentColor"
              height="1em"
              viewBox="0 0 256 256"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <rect
                fill="none"
                height="256"
                width="256"
              />
              <polyline
                fill="none"
                points="174 86 216 128 174 170"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="16"
              />
              <line
                fill="none"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="16"
                x1="104"
                x2="216"
                y1="128"
                y2="128"
              />
              <path
                d="M104,216H48a8,8,0,0,1-8-8V48a8,8,0,0,1,8-8h56"
                fill="none"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="16"
              />
            </svg>
             
            Fazer logout
          </a>
        </div>
      </div>
    </div>
    <button
      aria-label="button-toggle-sidebar"
      data-cy="button-cart"
    >
      <svg
        fill="currentColor"
        height="1em"
        viewBox="0 0 256 256"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <rect
          fill="none"
          height="256"
          width="256"
        />
        <circle
          cx="80"
          cy="216"
          r="16"
        />
        <circle
          cx="184"
          cy="216"
          r="16"
        />
        <path
          d="M42.3,72H221.7l-26.4,92.4A15.9,15.9,0,0,1,179.9,176H84.1a15.9,15.9,0,0,1-15.4-11.6L32.5,37.8A8,8,0,0,0,24.8,32H8"
          fill="none"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="16"
        />
      </svg>
      <div
        class="c-egqsFx"
      >
        1
      </div>
    </button>
  </div>
</div>
`;
